yjit_codegen.c

```diff
@@ -689,6 +689,118 @@ gen_duparray(jitstate_t* jit, ctx_t* ctx)
     return YJIT_KEEP_COMPILING;
 }
 
+static void
+guard_object_is_heap(codeblock_t *cb, ctx_t *ctx, uint8_t *side_exit, x86opnd_t object_opnd)
+{
+    ADD_COMMENT(cb, "guard object is heap");
+
+    // Test that the object is not an immediate
+    test(cb, object_opnd, imm_opnd(RUBY_IMMEDIATE_MASK));
+    jnz_ptr(cb, side_exit);
+
+    // Test that the object is not false or nil
+    cmp(cb, object_opnd, imm_opnd(Qnil));
+    RUBY_ASSERT(Qfalse < Qnil);
+    jbe_ptr(cb, side_exit);
+}
+
+static inline void
+guard_object_is_array(codeblock_t *cb, x86opnd_t object_opnd, x86opnd_t flags_opnd, ctx_t *ctx, uint8_t *side_exit)
+{
+    guard_object_is_heap(cb, ctx, side_exit, object_opnd);
+    ADD_COMMENT(cb, "guard object is array");
+
+    // Pull out the type mask
+    mov(cb, flags_opnd, member_opnd(object_opnd, struct RBasic, flags));
+    and(cb, flags_opnd, imm_opnd(RUBY_T_MASK));
+
+    // Compare the result with T_ARRAY
+    cmp(cb, flags_opnd, imm_opnd(T_ARRAY));
+    jne_ptr(cb, side_exit);
+}
+
+// push enough nils onto the stack to fill out an array
+static codegen_status_t
+gen_expandarray(jitstate_t* jit, ctx_t* ctx)
+{
+    int flag = (int) jit_get_arg(jit, 1);
+
+    // If this instruction has the splat flag, then bail out.
+    if (flag & 0x01) {
+        GEN_COUNTER_INC(cb, expandarray_splat);
+        return YJIT_CANT_COMPILE;
+    }
+
+    // If this instruction has the postarg flag, then bail out.
+    if (flag & 0x02) {
+        GEN_COUNTER_INC(cb, expandarray_postarg);
+        return YJIT_CANT_COMPILE;
+    }
+
+    // num is the number of requested values. If there aren't enough in the
+    // array then we're going to push on nils.
+    rb_num_t num = (rb_num_t) jit_get_arg(jit, 0);
+
+    // If we don't actually want any values, then just return.
+    if (num == 0) {
+        return YJIT_KEEP_COMPILING;
+    }
+
+    uint8_t *side_exit = yjit_side_exit(jit, ctx);
+
+    // Move the array from the stack into REG0 and check that it's an array.
+    mov(cb, REG0, ctx_stack_pop(ctx, 1));
+    guard_object_is_array(cb, REG0, REG1, ctx, COUNTED_EXIT(side_exit, expandarray_not_array));
+
+    uint32_t embedded_length_label = cb_new_label(cb, "ea_embedded");
+    uint32_t push_values_label = cb_new_label(cb, "ea_push_vals");
+
+    // Pull out the embed flag to check if it's an embedded array.
+    mov(cb, REG1, member_opnd(REG0, struct RBasic, flags));
+    and(cb, REG1, imm_opnd(RARRAY_EMBED_FLAG));
+    jmp_label(cb, embedded_length_label);
+
+    // Pull out the length of the heap array and write it to REG1.
+    mov(cb, REG1, member_opnd(REG0, struct RArray, as.heap.len));
+    jmp_label(cb, push_values_label);
+
+    // Pull out the length of the embedded array and write it to REG1.
+    cb_write_label(cb, embedded_length_label);
+    mov(cb, REG1, member_opnd(REG0, struct RBasic, flags));
+    and(cb, REG1, imm_opnd(RARRAY_EMBED_LEN_MASK));
+    shr(cb, REG1, imm_opnd(RARRAY_EMBED_LEN_SHIFT));
```

Comment:
We might be able to do this with a conditional move as well.

In a WIP test branch I made this if it helps 🙂 (disclaimer: didn't benchmark `cmov` vs jumps)

```
static void
gen_get_array_len(codeblock_t *cb, x86opnd_t target_reg, x86opnd_t source_reg)
{
    x86opnd_t flags_opnd = member_opnd(source_reg, struct RBasic, flags);
    x86opnd_t array_heap_len = member_opnd(source_reg, struct RArray, as.heap.len);

    // todo: combine flags memory accesses with a scratch reg
    mov(cb, target_reg, flags_opnd);
    and(cb, target_reg, imm_opnd(RARRAY_EMBED_LEN_MASK));
    shr(cb, target_reg, imm_opnd(RARRAY_EMBED_LEN_SHIFT));
    test(cb, flags_opnd, imm_opnd(RARRAY_EMBED_FLAG));
    cmovz(cb, target_reg, array_heap_len);
}
```

Reply:
Yeah! I wasn't sure which would be better/faster tbh.